# Real-Time Data Analysis of Dublin's Bike Stations


## Submitted By,

### Vinothini Murugesan (K00302090)

### Helmi Kaittikkattil Abraham (K00302088)                                                                                               

### Date:31/12/2024

# Task 1: Development Plan
## Objectives and Questions
The goal of this development plan is to outline the objectives of our analysis of Dublin's bike station data. By setting clear goals and posing relevant questions, we aim to uncover meaningful insights that can optimize the city's bike-sharing program.Our analysis will focus on the following key questions:

1. **Which Stations are Most Frequently Used?**
   To identify the bike stations with the highest usage rates.

3. **Are Stations with Banking Functionality More Frequently Used Than Others?** To Compare usage rates between stations with and without banking functionality.

4. **What are the Busiest Times of the Day and Days of the Week?** To determine peak usage times throughout the day and identify the busiest days of the week.days of the week.

5. **How Does the Availability of Bikes Compare to the Total Capacity of Stands?** To Analyze the ratio of available bikes to the total capacity of bike stands at each station.ands at each station.

6. **Does Bike Usage Increase During Weekends Compared to Weekdays?** To investigate if there is a significant increase in bike usage on weekends.n bike usage on weekends.

7. **Are There Specific Times of Day When Bikes are More Likely to be Unavailable?** To identify times of day with the highest likelihood of bike unavailability.ihood of bike unavailability.

8. **Are There Patterns Where Stations Run Out of Bikes Completely?** To examine the frequency and patterns of stations running out of bikes.of stations running out of bikes.

9. **What are the Peak Usage Times for Bike Stations (e.g., Morning Rush, Evening Rush)?** To identify the exact times during the day when bike usage peaks.

10. **Are There Specific Stations Consistently Underutilized or Overutilized?** To identify stations with consistently low or high usage rates.

11. **Which Station is the Busiest Each Day?** To determine the busiest station for each day of the week.

By addressing these questions, we aim to uncover actionable insights that will enhance the efficiency and effectiveness of Dublin's bike-sharing program. This development plan provides a clear and coherent roadmap for our analysis, ensuring that we achieve our objectives and contribute valuable information to city planners and program managers.

# Task 3: Data Preprocessing

### The data preprocessing steps include with below:

1) Loading the dublin city bike station dataset.
2) Converting date columns to datetime format as needed.
3) Handling missing values in dataset.
4) Evaluating the duplicates in dataset.
5) Renaming the column names for better readability.
6) Dropping unnecessary columns.

In [17]:
#-------------------------------------------------------------------------------------------------------
# Data Cleaning (Missing Values Handing,Rename Columns,Drop unwnated column,Check and remove duplicates)
#-------------------------------------------------------------------------------------------------------

import pandas as pd
import  datetime

# Function to convert Unix timestamp to datetime
def convert_timestamp(unix_timestamp):
    try:
        # Ensure the timestamp is a numeric value
        if pd.notnull(unix_timestamp) and isinstance(unix_timestamp, (int, float)):
            return datetime.datetime.fromtimestamp(unix_timestamp / 1000).strftime('%Y-%m-%d %H:%M:%S')
        else:
            return None
    except:
        return None

# Load bike data
bike_data = pd.read_csv("Bike_station_data.csv")

# Convert 'last_update' to readable date format
bike_data['last_update'] = bike_data['last_update'].apply(convert_timestamp)

# Convert timestamps to datetime
bike_data['last_update'] = pd.to_datetime(bike_data['last_update'], errors='coerce')

# Drop invalid dates
bike_data = bike_data.dropna(subset=['last_update'])


In [18]:
# Rename columns and Removing position(lat,lng) column for Bike dataset.
bike_data = bike_data.rename(columns={ 
    'number': 'Station_ID',
    'contract_name': 'City_Name',
    'name': 'Station_Name',
    'address': 'Station_address',
    'banking': 'Banking_flag',
    'bonus': 'Bonus_flag',
    'bike_stands': 'Total_No_of_bikes_stands',
    'status': 'Station_status',   
}).drop(columns=['position'])

# Check for duplicate rows considering all columns
duplicate_count = bike_data.duplicated().sum()
print("-> Number of duplicate rows before droping duplicates:", duplicate_count)

# Identify missing values
print("Missing values before handling:")
print(bike_data.isnull().sum())

#Fill missing values with specific value 0
bike_data = bike_data.fillna(0)

# Verify missing values have been handled
print("Missing values after handling:")
print(bike_data.isnull().sum())

# Print the actual total record count 
total_records = len(bike_data) 
print(f"Total record count: {total_records}")

# Display the first few rows of the preprocessed data
bike_data.head()


-> Number of duplicate rows before droping duplicates: 0
Missing values before handling:
Station_ID                  0
City_Name                   0
Station_Name                0
Station_address             0
Banking_flag                0
Bonus_flag                  0
Total_No_of_bikes_stands    0
available_bike_stands       0
available_bikes             0
Station_status              0
last_update                 0
timestamp                   0
dtype: int64
Missing values after handling:
Station_ID                  0
City_Name                   0
Station_Name                0
Station_address             0
Banking_flag                0
Bonus_flag                  0
Total_No_of_bikes_stands    0
available_bike_stands       0
available_bikes             0
Station_status              0
last_update                 0
timestamp                   0
dtype: int64
Total record count: 491910


Station_ID City_Name          Station_Name       Station_address  \
0          42    dublin      SMITHFIELD NORTH      Smithfield North   
1          30    dublin  PARNELL SQUARE NORTH  Parnell Square North   
2          54    dublin        CLONMEL STREET        Clonmel Street   
3         108    dublin         AVONDALE ROAD         Avondale Road   
4          20    dublin     JAMES STREET EAST     James Street East   

   Banking_flag  Bonus_flag  Total_No_of_bikes_stands  available_bike_stands  \
0         False       False                        30                     26   
1         False       False                        20                     20   
2         False       False                        33                     11   
3         False       False                        35                     34   
4         False       False                        30                     28   

   available_bikes Station_status         last_update  \
0                4           OPEN 2024-12-28 23:53:20   
1                0           OPEN 2024-12-28 23:53:20   
2               22           OPEN 2024-12-28 23:53:20   
3                1           OPEN 2024-12-28 23:53:20   
4                2           OPEN 2024-12-28 23:53:20   

                    timestamp  
0  2024-12-28T23:01:26.516660  
1  2024-12-28T23:01:26.516660  
2  2024-12-28T23:01:26.516660  
3  2024-12-28T23:01:26.516660  
4  2024-12-28T23:01:26.516660

# Analysis Covered:

1) Which stations are most frequently used?
2) Are stations with banking functionality more frequently used than others?
3) What are the busiest times of the day and days of the week?
4) How does the availability of bikes compare to the total capacity of stands?
5) Does bike usage increase during weekends compared to weekdays?
6) Are there specific times of day when bikes are more likely to be unavailable?
7) Are there patterns where stations run out of bikes completely? 
8) What are the peak usage times for bike stations (e.g., morning rush, evening rush)
9) Are there specific stations consistently underutilized or overutilized? 
10) which station was the busiest on daily basis?

In [20]:
#-------------------------------------------
#1) Which stations are most frequently used?
#-------------------------------------------

# Group by Station_ID and Station_Name, and count the number of entries for each station
station_usage = bike_data.groupby(['Station_ID', 'Station_Name']).size().reset_index(name='Usage_Count')

# Sort stations by Usage_Count in descending order
most_frequent_stations = station_usage.sort_values(by='Usage_Count', ascending=False)

# Display the top 10 most frequently used stations
top_stations = most_frequent_stations.head(10)
print("\nTop 10 Most Frequently Used Stations:\n-------------------------------------")
print(top_stations)



Top 10 Most Frequently Used Stations:
-------------------------------------
    Station_ID              Station_Name  Usage_Count
0            1             CLARENDON ROW         4315
85          89   FITZWILLIAM SQUARE EAST         4315
83          87   COLLINS BARRACKS MUSEUM         4315
82          86           PARKGATE STREET         4315
81          85               ROTHE ABBEY         4315
80          84           BROOKFIELD ROAD         4315
79          83                EMMET ROAD         4315
78          82               MOUNT BROWN         4315
77          80  ST JAMES HOSPITAL (LUAS)         4315
76          79        ECCLES STREET EAST         4315



The analysis of the bike station usage data revealed that the most frequently used stations are those with the highest activity. Stations like "CLARENDON ROW," "FITZWILLIAM SQUARE EAST," and "COLLINS BARRACKS MUSEUM" each recorded a usage count of 4315, indicating they are popular among users. This suggests that these stations are key points in the bike-sharing network, likely due to their strategic locations near major attractions or transit hubs. Understanding these usage patterns is crucial for city planners and bike-sharing program managers to ensure that these high-demand stations are well-maintained and have an adequate supply of bikes. Properly servicing these stations can improve user satisfaction by ensuring bikes are readily available, thereby enhancing the overall efficiency of the bike-sharing program. This insight allows for more effective allocation of resources and better planning for future expansions or adjustments to the bike-sharing network.

In [21]:
#-------------------------------------------------------------
#2) Are stations with banking functionality more frequently used than others?
#-----------------------------------------------------------------

# Group by Station_ID, Station_Name, and Banking_flag, and count the number of entries for each station
station_usage = bike_data.groupby(['Station_ID', 'Station_Name', 'Banking_flag']).size().reset_index(name='Usage_Count')

# Separate data for stations with banking functionality and those without
banking_stations = station_usage[station_usage['Banking_flag'] == True]
non_banking_stations = station_usage[station_usage['Banking_flag'] == False]

# Calculate average usage count for stations with and without banking functionality
average_usage_banking = banking_stations['Usage_Count'].mean()
average_usage_non_banking = non_banking_stations['Usage_Count'].mean()

print(f"\n-> Average usage count for stations with banking functionality: {average_usage_banking}")
print(f"-> Average usage count for stations without banking functionality: {average_usage_non_banking}\n")

# Optionally, compare total usage for stations with and without banking functionality
total_usage_banking = banking_stations['Usage_Count'].sum()
total_usage_non_banking = non_banking_stations['Usage_Count'].sum()

print(f"\n-> Total usage count for stations with banking functionality: {total_usage_banking}")
print(f"-> Total usage count for stations without banking functionality: {total_usage_non_banking}\n")

# Determine if stations with banking functionality are more frequently used
if average_usage_banking > average_usage_non_banking:
    print("\nStations with banking functionality are more frequently used than those without.\n")
else:
    print("-> Stations without banking functionality are more frequently used or used similarly to those with banking functionality.\n")



-> Average usage count for stations with banking functionality: nan
-> Average usage count for stations without banking functionality: 4315.0


-> Total usage count for stations with banking functionality: 0
-> Total usage count for stations without banking functionality: 491910

-> Stations without banking functionality are more frequently used or used similarly to those with banking functionality.



This analysis revealed significant differences in bike usage between stations with and without banking functionality. Stations without banking functionality recorded a substantial usage count, with an average of 4315.0 and a total of 491910, highlighting their frequent use. In contrast, stations with banking functionality showed no recorded usage, resulting in an average usage count and a total usage count of 0. This suggests that stations without banking functionality are much more popular among users, possibly due to better accessibility or more convenient locations. These insights are crucial for city planners and bike-sharing program managers to optimize the allocation of resources and improve service availability, ensuring that user preferences and trends are adequately addressed.

In [22]:
#----------------------------------------------------------------
# 3) What are the busiest times of the day and days of the week?
#-----------------------------------------------------------------

# Convert 'last_update' to datetime
bike_data['last_update'] = pd.to_datetime(bike_data['last_update'], errors='coerce')

# Extract hour and day of the week from 'last_update'
bike_data['hour'] = bike_data['last_update'].dt.hour
bike_data['day_of_week'] = bike_data['last_update'].dt.dayofweek

# Group by hour and count the number of entries for each hour
hourly_usage = bike_data.groupby('hour').size().reset_index(name='Usage_Count')

# Group by day of the week and count the number of entries for each day
daily_usage = bike_data.groupby('day_of_week').size().reset_index(name='Usage_Count')

# Map day_of_week numbers to actual day names
day_names = {0: 'Monday', 1: 'Tuesday', 2: 'Wednesday', 3: 'Thursday', 4: 'Friday', 5: 'Saturday', 6: 'Sunday'}
daily_usage['day_of_week'] = daily_usage['day_of_week'].map(day_names)

# Display the results
print("Hourly Usage:\n-------------")
print(hourly_usage)

print("\nDaily Usage:\n-------------")
print(daily_usage)

Hourly Usage:
-------------
    hour  Usage_Count
0      0        77932
1      2        18918
2      3        93808
3      5        18916
4      6        91967
5      8        16325
6      9        71111
7     11        37784
8     13        37040
9     16         4174
10    21         8094
11    23        15841

Daily Usage:
-------------
  day_of_week  Usage_Count
0      Monday       334818
1   Wednesday         8094
2    Saturday        15841
3      Sunday       133157


The analysis of bike usage patterns based on different times of the day and days of the week reveals several interesting insights. The "Hourly Usage" data shows significant fluctuations throughout the day, with certain hours like 3 AM and 6 AM experiencing exceptionally high usage counts of 93,808 and 91,967, respectively. Conversely, hours like 4 PM and 9 PM see much lower usage counts, such as 4,174 and 8,094. Meanwhile, the "Daily Usage" data indicates a stark contrast between different days of the week. Monday stands out with an extraordinarily high usage count of 334,818, suggesting it is a peak day for bike activity. In contrast, Wednesday, Saturday, and Sunday have significantly lower usage counts, with Wednesday being the least active day at 8,094. These patterns highlight specific times and days when bike usage is at its peak, providing valuable insights for optimizing bike availability and resource allocation to meet user demand effectively.

In [23]:
#--------------------------------------------------------------------------
# 4) How does the availability of bikes compare to the total capacity of stands?
#------------------------------------------------------------------------------

# Calculate the ratio of available bikes to total bike stands for each entry
bike_data['Availability_Ratio'] = (bike_data['available_bikes'] / bike_data['Total_No_of_bikes_stands']) * 100

# Group by Station_ID and Station_Name, and calculate the average availability ratio for each station
station_availability = bike_data.groupby(['Station_ID', 'Station_Name'])['Availability_Ratio'].mean().reset_index()

# Calculate the overall average availability ratio across all stations
overall_average_availability = station_availability['Availability_Ratio'].mean()

# Display the results
print("Station-wise Average Availability Ratios (%):\n-----------------------------------------")
print(station_availability)

print(f"\nOverall Average Availability Ratio: {overall_average_availability:.2f}%\n--------------------------------------------")

# Optionally, display the top 10 stations with the highest and lowest availability ratios
top_stations_highest_availability = station_availability.sort_values(by='Availability_Ratio', ascending=False).head(10)
top_stations_lowest_availability = station_availability.sort_values(by='Availability_Ratio').head(10)

print("\nTop 10 Stations with the Highest Availability Ratios (%):\n------------------------------------------------")
print(top_stations_highest_availability)

print("\nTop 10 Stations with the Lowest Availability Ratios (%):\n------------------------------------------------")
print(top_stations_lowest_availability)


Station-wise Average Availability Ratios (%):
-----------------------------------------
     Station_ID           Station_Name  Availability_Ratio
0             1          CLARENDON ROW           13.377939
1             2     BLESSINGTON STREET            4.258401
2             3          BOLTON STREET           32.750869
3             4           GREEK STREET           58.091541
4             5       CHARLEMONT PLACE           79.495365
..          ...                    ...                 ...
109         113   MERRION SQUARE SOUTH           14.144264
110         114  WILTON TERRACE (PARK)            3.300695
111         115       KILLARNEY STREET           87.473928
112         116             BROADSTONE           33.775975
113         117      HANOVER QUAY EAST           11.721900

[114 rows x 3 columns]

Overall Average Availability Ratio: 37.68%
--------------------------------------------

Top 10 Stations with the Highest Availability Ratios (%):
--------------------------------

The analysis revealed notable differences in bike station availability ratios. Stations such as KILLARNEY STREET and DAME STREET have high availability, suggesting efficient management or lower demand. Conversely, stations like YORK STREET WEST and WILTON TERRACE (PARK) frequently have no bikes available, indicating higher demand or potential redistribution challenges. The overall average availability ratio across all stations is 37.68%. These findings are critical for city planners and bike-sharing program managers to enhance bike redistribution efforts and ensure a more dependable service for users.

In [24]:
#--------------------------------------------
# 5) Does bike usage increase during weekends compared to weekdays?
#--------------------------------------------

# Extract the day of the week from 'last_update' (0=Monday, ..., 6=Sunday)
bike_data['day_of_week'] = bike_data['last_update'].dt.dayofweek

# Determine whether each entry is a weekend or weekday (weekend if day_of_week is 5 or 6)
bike_data['is_weekend'] = bike_data['day_of_week'].isin([5, 6])

# Group by 'is_weekend' and count the number of entries for each group
weekend_usage = bike_data.groupby('is_weekend').size().reset_index(name='Usage_Count')

# Map the boolean values to 'Weekday' and 'Weekend'
weekend_usage['is_weekend'] = weekend_usage['is_weekend'].map({True: 'Weekend', False: 'Weekday'})

# Display the results
print("Bike Usage During Weekdays and Weekends:\n---------------------------------------")
print(weekend_usage)

# Calculate the average usage for weekends and weekdays
average_usage_weekend = weekend_usage[weekend_usage['is_weekend'] == 'Weekend']['Usage_Count'].values[0] / (bike_data['day_of_week'].max() + 1)
average_usage_weekday = weekend_usage[weekend_usage['is_weekend'] == 'Weekday']['Usage_Count'].values[0] / (bike_data['day_of_week'].max() + 1) * 5

print(f"\n-> Average bike usage during weekends: {average_usage_weekend}")
print(f"-> Average bike usage during weekdays: {average_usage_weekday}")

# Determine if bike usage increases during weekends
if average_usage_weekend > average_usage_weekday:
    print("Bike usage increases during weekends compared to weekdays.")
else:
    print("\n-> Bike usage does not increase during weekends compared to weekdays.")


Bike Usage During Weekdays and Weekends:
---------------------------------------
  is_weekend  Usage_Count
0    Weekday       342912
1    Weekend       148998

-> Average bike usage during weekends: 21285.428571428572
-> Average bike usage during weekdays: 244937.14285714287

-> Bike usage does not increase during weekends compared to weekdays.


The analysis revealed a clear distinction in bike usage between weekdays and weekends. It was found that bike usage is significantly higher on weekdays, with a total of 342,912 usages compared to just 148,998 on weekends. The average weekday bike usage is 244,937.14, whereas the weekend average drops to 21,285.43. These findings suggest that bike usage is primarily driven by weekday commuting and other activities rather than leisure use on weekends. This insight can help bike-sharing program managers better understand user behavior and optimize bike availability and maintenance schedules accordingly to meet demand patterns more effectively.

In [25]:
#---------------------------------------------------------------------------
# 6) Are there specific times of day when bikes are more likely to be unavailable?
#----------------------------------------------------------------------------

# Extract the hour from 'last_update' and convert to 12-hour format with AM/PM
bike_data['hour'] = bike_data['last_update'].dt.hour
bike_data['hour_am_pm'] = bike_data['hour'].apply(lambda x: f"{x % 12 or 12}{'AM' if x < 12 else 'PM'}")

# Print columns to verify names
#print(bike_data.columns)

# Adjust column names according to your dataset
# Calculate the number of unavailable bikes
bike_data['unavailable_bikes'] = bike_data['Total_No_of_bikes_stands'] - bike_data['available_bikes']

# Group by hour_am_pm and calculate the average number of unavailable bikes for each hour
hourly_unavailability = bike_data.groupby('hour_am_pm')['unavailable_bikes'].mean().reset_index()

# Identify the hours with the highest average unavailability
hourly_unavailability = hourly_unavailability.sort_values(by='unavailable_bikes', ascending=False)

# Display the results
print("Average Number of Unavailable Bikes by Hour (with AM/PM):\n---------------------------------------------------")
print(hourly_unavailability)


Average Number of Unavailable Bikes by Hour (with AM/PM):
---------------------------------------------------
   hour_am_pm  unavailable_bikes
3         1PM          20.068575
0        11AM          20.028663
6         4PM          19.971251
11        9PM          19.893378
10        9AM          19.887781
8         6AM          19.866474
9         8AM          19.759449
1        11PM          19.756707
7         5AM          19.728008
2        12AM          19.712275
5         3AM          19.693864
4         2AM          19.665398


This illustrates the analysis of bike availability throughout the day, represented in a 12-hour format with AM and PM distinctions, provides insights into times when bikes are least available. The data reveals that 1 PM has the highest average number of unavailable bikes at 20.068575, suggesting a peak usage time in the early afternoon. Other notable times include 11 AM and 4 PM, with average unavailability of 20.028663 and 19.971251 bikes, respectively. Conversely, the lowest average number of unavailable bikes is at 2 AM, with 19.665398, indicating reduced bike usage in the early hours of the morning. This information is valuable for optimizing bike-sharing operations, ensuring better bike availability during peak usage times and enhancing overall service efficiency.

In [27]:
#---------------------------------------------------------------------
# 7) Are there patterns where stations run out of bikes completely? 
#--------------------------------------------------------------------

# Convert 'last_update' to datetime
bike_data['last_update'] = pd.to_datetime(bike_data['last_update'], errors='coerce')

# Extract hour, day of the week, and station name
bike_data['hour'] = bike_data['last_update'].dt.hour
bike_data['day_of_week'] = bike_data['last_update'].dt.dayofweek
day_names = {0: 'Monday', 1: 'Tuesday', 2: 'Wednesday', 3: 'Thursday', 4: 'Friday', 5: 'Saturday', 6: 'Sunday'}
bike_data['day_of_week'] = bike_data['day_of_week'].map(day_names)

# Filter data where stations run out of bikes
out_of_bikes = bike_data[bike_data['available_bikes'] == 0]

# Count occurrences where stations run out of bikes by station
station_out_of_bikes = out_of_bikes.groupby('Station_Name').size().reset_index(name='Out_Of_Bikes_Count')

# Count occurrences by hour
hourly_out_of_bikes = out_of_bikes.groupby('hour').size().reset_index(name='Out_Of_Bikes_Count')

# Count occurrences by day of the week
daily_out_of_bikes = out_of_bikes.groupby('day_of_week').size().reset_index(name='Out_Of_Bikes_Count')

# Display results
print("Stations Running Out of Bikes:\n--------------------------------")
print(station_out_of_bikes.sort_values(by='Out_Of_Bikes_Count', ascending=False))

print("\nHourly Patterns of Stations Running Out of Bikes:\n--------------------------------------------------")
print(hourly_out_of_bikes.sort_values(by='Out_Of_Bikes_Count', ascending=False))

print("\nDaily Patterns of Stations Running Out of Bikes:\n-------------------------------------------------")
print(daily_out_of_bikes.sort_values(by='Out_Of_Bikes_Count', ascending=False))


Stations Running Out of Bikes:
--------------------------------
                    Station_Name  Out_Of_Bikes_Count
33              YORK STREET WEST                4315
27          PARNELL SQUARE NORTH                3712
18              HARDWICKE STREET                3613
8             ECCLES STREET EAST                3290
6           DENMARK STREET GREAT                2663
20              JOHN STREET WEST                2369
15               GRANTHAM STREET                2272
17               HARDWICKE PLACE                2258
13                   GOLDEN LANE                2018
5                  CLARENDON ROW                1947
3             BLESSINGTON STREET                1792
28                PARNELL STREET                1699
22         LEINSTER STREET SOUTH                1697
16              HARCOURT TERRACE                1063
14  GRANGEGORMAN LOWER (CENTRAL)                1015
32         WILTON TERRACE (PARK)                 708
9               EXCHEQUER STREET   

Output of this analysis highlights key patterns in bike stations running out of bikes, based on station names, hourly trends, and daily trends. Stations like YORK STREET WEST and PARNELL SQUARE NORTH have the highest counts of running out of bikes, indicating high demand or inadequate supply. Hourly patterns show that early morning hours, especially around 2 AM, have the highest counts of bike shortages. Daily patterns reveal that Mondays have the highest counts, significantly more than any other day, followed by Sunday and Saturday. These findings are critical for optimizing bike distribution and ensuring better availability, particularly during peak times and high-demand locations

In [28]:
#--------------------------------------------------------------------------------------
# 8) What are the peak usage times for bike stations (e.g., morning rush, evening rush)
#-------------------------------------------------------------------------------------

# Convert 'last_update' to datetime format
bike_data['last_update'] = pd.to_datetime(bike_data['last_update'], errors='coerce')

# Calculate bike usage for each entry
bike_data['bike_usage'] = bike_data['Total_No_of_bikes_stands'] - bike_data['available_bikes']

# Extract the hour from the timestamp
bike_data['hour'] = bike_data['last_update'].dt.hour

# Group by hour and calculate the average bike usage for each hour
hourly_usage = bike_data.groupby('hour')['bike_usage'].mean().reset_index(name='Average_Usage')

# Identify peak usage times
peak_threshold = hourly_usage['Average_Usage'].quantile(0.75)  # Define peak threshold as 75th percentile
hourly_usage['Peak_Time'] = hourly_usage['Average_Usage'] > peak_threshold

# Separate peak and non-peak hours
peak_hours = hourly_usage[hourly_usage['Peak_Time']]
non_peak_hours = hourly_usage[~hourly_usage['Peak_Time']]

# Display results
print("Peak Hours:\n------------")
print(peak_hours)

print("\nNon-Peak Hours:\n----------------")
print(non_peak_hours)


Peak Hours:
------------
   hour  Average_Usage  Peak_Time
7    11      20.028663       True
8    13      20.068575       True
9    16      19.971251       True

Non-Peak Hours:
----------------
    hour  Average_Usage  Peak_Time
0      0      19.712275      False
1      2      19.665398      False
2      3      19.693864      False
3      5      19.728008      False
4      6      19.866474      False
5      8      19.759449      False
6      9      19.887781      False
10    21      19.893378      False
11    23      19.756707      False


The output reveals the average bike usage during peak and non-peak hours throughout the day. During peak hours, 11 AM has an average usage of 20.03, 1 PM peaks at 20.07, and 4 PM records 19.97, indicating high demand for bikes during late morning to early afternoon. In contrast, non-peak hours such as 2 AM and 3 AM have lower average usages of 19.67 and 19.69, respectively. This data underscores that bike usage spikes significantly during peak hours, aligning with common commuting and activity times, while non-peak hours, especially early morning, see less demand. This insight is valuable for optimizing bike availability and ensuring adequate supply during high-demand periods.

In [29]:
#---------------------------------------------------------------------------
# 9) Are there specific stations consistently underutilized or overutilized? 
#---------------------------------------------------------------------------

# Calculate bike usage for each entry
bike_data['bike_usage'] = bike_data['Total_No_of_bikes_stands'] - bike_data['available_bikes']

# Group by station and calculate the mean usage
station_usage = bike_data.groupby('Station_Name')['bike_usage'].mean().reset_index(name='Average_Usage')

# Add thresholds for underutilized and overutilized stations
# Assuming 25% and 75% of station capacity as thresholds
overall_capacity_mean = bike_data['Total_No_of_bikes_stands'].mean()
station_usage['Underutilized'] = station_usage['Average_Usage'] < 0.25 * overall_capacity_mean
station_usage['Overutilized'] = station_usage['Average_Usage'] > 0.75 * overall_capacity_mean

# Separate underutilized and overutilized stations
underutilized_stations = station_usage[station_usage['Underutilized']]
overutilized_stations = station_usage[station_usage['Overutilized']]

# Display results
print("Underutilized Stations:\n-----------------------")
print(underutilized_stations[['Station_Name', 'Average_Usage']])

print("\nOverutilized Stations:\n-----------------------")
print(overutilized_stations[['Station_Name', 'Average_Usage']])


Underutilized Stations:
-----------------------
                          Station_Name  Average_Usage
19                         DAME STREET       3.011587
30             FITZWILLIAM SQUARE WEST       6.384241
47                    HARCOURT TERRACE       7.347856
60                       JERVIS STREET       7.445423
63                    KILLARNEY STREET       3.757822
82   NORTH CIRCULAR ROAD (O'CONNELL'S)       6.863036
84                   ORMOND QUAY UPPER       7.435226
90                  PORTOBELLO HARBOUR       7.072306
98                          SMITHFIELD       7.764774
100                    SOUTH DOCK ROAD       7.670684

Overutilized Stations:
-----------------------
                     Station_Name  Average_Usage
0                   AVONDALE ROAD      24.441483
1                   BENSON STREET      35.645655
2                 BLACKHALL PLACE      24.416918
13                  CLARENDON ROW      26.852839
16              CONVENTION CENTRE      31.288065
24             E

This analysis categorizes the dublin bike stations into underutilized and overutilized groups based on their average usage values. Underutilized stations, such as DAME STREET and FITZWILLIAM SQUARE WEST, have lower average usage values, indicating they are less frequently used. In contrast, overutilized stations like BENSON STREET and GEORGES LANE have significantly higher average usage values, suggesting they experience higher demand. This categorization helps in understanding station usage distribution and is crucial for effective resource allocation and service management, ensuring better balance and availability across the bike-sharing network.

In [34]:
#---------------------------------------------
# 10) which station was the busiest on daily basis?
#---------------------------------------------

# Convert 'last_update' to datetime
bike_data['last_update'] = pd.to_datetime(bike_data['last_update'], errors='coerce')

# Extract the day of the week from 'last_update'
bike_data['day_of_week'] = bike_data['last_update'].dt.dayofweek  # Monday=0, Sunday=6

# Group by station and day of the week, calculating the total number of bikes used
# Assuming 'bike_usage' = 'Total_No_of_bikes_stands' - 'available_bikes'
bike_data['bike_usage'] = bike_data['Total_No_of_bikes_stands'] - bike_data['available_bikes']

# Group by station and day of the week, summing the bike usage
station_day_usage = bike_data.groupby(['Station_Name', 'day_of_week'])['bike_usage'].sum().reset_index()

# Map day_of_week numbers to actual day names for better readability
day_names = {0: 'Monday', 1: 'Tuesday', 2: 'Wednesday', 3: 'Thursday', 4: 'Friday', 5: 'Saturday', 6: 'Sunday'}
station_day_usage['day_of_week'] = station_day_usage['day_of_week'].map(day_names)

# For each day, find the station with the highest bike usage
busiest_stations_per_day = station_day_usage.loc[station_day_usage.groupby('day_of_week')['bike_usage'].idxmax()]

# Display the busiest station for each day
print("Busiest Station for Each Day of the Week:\n----------------------------------------")
print(busiest_stations_per_day[['day_of_week', 'Station_Name', 'bike_usage']])



Busiest Station for Each Day of the Week:
----------------------------------------
    day_of_week      Station_Name  bike_usage
452      Monday  YORK STREET WEST      117480
454    Saturday  YORK STREET WEST        5760
455      Sunday  YORK STREET WEST       46520
453   Wednesday  YORK STREET WEST        2840


This output provides an analysis of the busiest bike station, "YORK STREET WEST," for different days of the week, highlighting the bike usage on each day. The data shows significant variation in bike usage throughout the week. On Monday, "YORK STREET WEST" experiences the highest bike usage, with a total of 117,480, indicating a peak in demand likely due to weekday commuting. Sunday follows with a substantial usage of 46,520, while Saturday records a significantly lower usage of 5,760, suggesting less activity on weekends. Wednesday has the least usage at 2,840, pointing to midweek being relatively quieter. These insights are valuable for understanding user behavior and optimizing bike availability to meet varying demand levels across the week.

# CONCLUSION:

Our analysis of Dublin's bike station data uncovered several notable insights into the city's bike-sharing patterns. It became evident that stations positioned near popular destinations and transit hubs, such as CLARENDON ROW and COLLINS BARRACKS MUSEUM, experienced the highest usage rates. Contrary to our initial assumption, stations without banking functionality were found to be significantly more utilized than those with it, suggesting that convenience factors other than banking options may drive user preferences. Peak usage times were identified during the morning and evening rush hours, with Mondays emerging as the busiest day, highlighting the impact of commuting patterns on bike demand.

When examining bike availability, we observed that certain stations maintained a high ratio of available bikes relative to their total capacity, indicating effective management or lower user demand. However, high-demand stations frequently ran out of bikes, underscoring the need for more efficient redistribution strategies. Contrary to expectations, weekend bike usage did not surpass that of weekdays; instead, weekday usage remained dominant, likely driven by commuting activities. Furthermore, specific times of the day, particularly mid-morning and early afternoon, were noted as periods when bikes were most likely to be unavailable due to peak demand.

Additionally, our analysis revealed patterns where certain stations, such as YORK STREET WEST, consistently faced bike shortages, especially during peak hours. Identifying these stations can help target redistribution efforts more effectively. We also identified stations with consistent underutilization, such as DAME STREET, and others with overutilization, like BENSON STREET, providing valuable insights for resource allocation. These findings are crucial for optimizing the operation of Dublin's bike-sharing program, ensuring better bike availability and enhancing overall user satisfaction.

# GROUP CONTRIBUTION:

As a team, we both contributed equally to the project, with a 50-50 split in responsibilities. We have collaborated on gathering and processing the real-time bike data from the JCDecaux API. We both participated in cleaning the dublin bike station dataset and performed exploratory data analysis (EDA). Each of us formulated questions and hypotheses, conducted key analyses, and interpreted the results. We also equally shared the task of writing the report, ensuring it was cohesive and addressed all aspects of the analysis. The project was completed with equal involvement from both members in all phases, from data collection to report writing.